skfuzzy/control/controlsystem.py

```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        self.ctrl = ctrl
+        self._cache = []
+        self._cahced_graph = None
+
+
+    def __iter__(self):
+        # Determine if we can return the cached version or must calc new
+        if self._cahced_graph is not self.ctrl.graph:
+            # The controller is still using a different version of the graph
+            #  than we created the rule order for.  Thus, make new cache
+            self._init_state()
+            self._cache = list(self._process_rules(self.all_rules[:]))
+            self._cahced_graph = self.ctrl.graph
+
+        for n, r in enumerate(self._cache):
+            yield r
+        assert n == len(self.all_rules)-1, "Not all rules exposed"
+
+    def _init_state(self):
+        # This graph will represent what's been calculated so far.  We
+        # initialize it to just the antecedents as they, by definition, already
+        # have fuzzy values
+        self.calced_graph = nx.DiGraph()
+        for a in self.ctrl.antecedents:
+            self.calced_graph.add_star([a, ] + a.terms.values())
+
+        self.all_graph = self.ctrl.graph
+
+        self.all_rules = []
+        for node in self.all_graph.nodes():
+            if isinstance(node, Rule):
+                self.all_rules.append(node)
+
+    def _process_rules(self, rules):
+        # Recursive funcion to process rules in the correct firing order
+        len_rules = len(rules)
+        skipped_rules = []
+        while len(rules) > 0:
+            rule = rules.pop(0)
+            if self._can_calc_rule(rule):
+                yield rule
+                # Add rule to calced graph
+                self.calced_graph = nx.compose(self.calced_graph, rule.graph)
+            else:
+                # We have not calculated the predecsors for this rule yet.
+                #  Skip it for now
+                skipped_rules.append(rule)
+
+        if len(skipped_rules) == 0:
+            # All done!
+            raise StopIteration()
+        else:
+            if len(skipped_rules) == len_rules:
+                # Avoid being caught in an infinite loop
+                raise RuntimeError("Unable to resolve rule execution order")
+            else:
+                # Recurse across the skipped rules
+                for r in self._process_rules(skipped_rules):
+                    yield r
+
+    def _can_calc_rule(self, rule):
+        # Check that we've exposed all inputs to this rule by ensuring
+        # the predecessor-degree of each predecessor node is the same
+        # in both the calculation graph and overall graph
+        for p in self.all_graph.predecessors_iter(rule):
+            assert isinstance(p, FuzzyVariableTerm)
+            if p not in self.calced_graph:
+                return False
+
+            all_degree = len(self.all_graph.predecessors(p))
+            calced_degree = len(self.calced_graph.predecessors(p))
+            if all_degree != calced_degree:
+                return False
+        return True
```

Comment:
Git changelogs are cleaner if every files ends with exactly one newline. It's because the newline is actually a hidden `\n` on the prior line, so any extensions to the file "change" the previous final line in the diff unless it was there before.